AdvLoggerPkg: Add Dxe AdvancedLoggerLib Unit Test

VivianNK opened the pull request:
## Description

Add AdvancedLoggerDxeLib Unit Test
Add mock for AdvancedLogger Protocol

- [ ] Impacts functionality?
- [ ] Impacts security?
- [ ] Breaking change?
- [x] Includes tests?
- [ ] Includes documentation?

## How This Was Tested

Local CI and building the unit test

## Integration Instructions

N/A

VivianNK committed:
AdvLoggerPkg: Add MockAdvancedLogger Protocol

VivianNK committed:
AdvLoggerPkg: Add AdvancedLoggerDxeLibGoogleTest

kuqin12 commented on AdvLoggerPkg/Library/AdvancedLoggerLib/Dxe/GoogleTest/AdvancedLoggerDxeLibGoogleTest.inf:
```diff
@@ -0,0 +1,45 @@
+  ../AdvancedLoggerLib.c # Source for extern mInitialized
+  AdvancedLoggerDxeLibGoogleTest.cpp
+  ../../../../Test/Mock/Library/GoogleTest/Protocol/MockAdvancedLogger.cpp
+  # MdePkg/Test/Mock/Library/GoogleTest/Protocol/MockDebugPort.cpp # Waiting on MockDebugPort
```
Does this line really have to be here?

VivianNK replied:
No it's not needed. There are a few lines that are commented out because they're waiting on the DebugProtocolMock to be in MU_BASECORE (but it is going into edk2 first).

kuqin12 commented on AdvLoggerPkg/Library/AdvancedLoggerLib/Dxe/GoogleTest/AdvancedLoggerDxeLibGoogleTest.inf:
```diff
@@ -0,0 +1,45 @@
+[Sources]
+  ../AdvancedLoggerLib.c # Source for extern mInitialized
+  AdvancedLoggerDxeLibGoogleTest.cpp
+  ../../../../Test/Mock/Library/GoogleTest/Protocol/MockAdvancedLogger.cpp
```
Would it be better if we make this protocol instance a library? Or does it really make sense to break out the protocol implementation into its own file?

VivianNK replied:
Do you mean make the mock of the protocol [AdvLoggerProtocol](https://github.com/microsoft/mu_plus/blob/release/202311/AdvLoggerPkg/Include/Protocol/AdvancedLogger.h) a mock library?

VivianNK replied:
Discussed offline. I moved the mock protocol function definitions and instantiation to the header file.

kuqin12 commented on AdvLoggerPkg/Library/AdvancedLoggerLib/Dxe/GoogleTest/AdvancedLoggerDxeLibGoogleTest.cpp:
```diff
@@ -0,0 +1,297 @@
+  { 0x434f695c, 0xef26, 0x4a12, { 0x9e, 0xba, 0xdd, 0xef, 0x00, 0x97, 0x49, 0x7c }
+  };
+
+  virtual void
```
For my learning purpose, what does `virtual` do here?

VivianNK replied:
`virtual` expects the function to be overridden by derived classes, so this keyword was mistakenly copied from the Test class definition. I fixed it so it uses "override" instead.

VivianNK committed:
AdvLoggerPkg: Fix GCC type error and update test

Add variable instead of const 0
From code change recently merged, the invalid signature test now needs
the DebugPortProtocolMock.

VivianNK committed:
AdvLoggerPkg: Remove MockAdvancedLogger.cpp

Delete MockAdvancedLogger.cpp and move mock function definitions
to MockAdvancedLogger.h.

VivianNK committed:
to squash: misc fixes

kuqin12 commented on AdvLoggerPkg/Test/Mock/Include/GoogleTest/Protocol/MockAdvancedLogger.h:
```diff
@@ -0,0 +1,55 @@
+};
+
+extern "C" {
+  extern ADVANCED_LOGGER_PROTOCOL  *gALProtocol;
```
Do you still need this? It just materialized below?
